### PR TITLE
[RDY] Add LTO support

### PR DESCRIPTION
**Pull request by zhou13**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,7 +2,7 @@
 #   intro: https://codingnest.com/basic-cmake/
 #   best practices (3.0+): https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1
 
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.9)
```
Comment (justinmk) [inline CMakeLists.txt]: Is there a way to use 2.8.12? We need to be able to build on as many systems as possible.

**Comment by zhou13** [comment]
@justinmk Updated.

**Comment by justinmk** [comment]
We tried LTO in https://github.com/neovim/neovim/pull/4313 , later reverted in https://github.com/mgraczyk/neovim/commit/c2b2e0e0f428f2146150a558f7051e4db6a1a2c4 .

But it looks like you did something like what was discussed in https://github.com/neovim/neovim/pull/4313#issuecomment-186923297 , which was never tried. 

@mhinz @Yamakaky @ZyX-I @jamessan @jszakmeister any comments/objections?

File: src/nvim/CMakeLists.txt [hunk 2]
```diff
@@ -399,6 +399,10 @@ if(JEMALLOC_FOUND)
   list(APPEND NVIM_EXEC_LINK_LIBRARIES ${JEMALLOC_LIBRARIES})
 endif()
```
Comment (jamessan) [inline src/nvim/CMakeLists.txt]: Please also add a `USE_LTO` option (which defaults to ON), so that LTO can easily be avoided.

Comment (zhou13) [reply src/nvim/CMakeLists.txt]: Done

**Comment by zhou13** [comment]
The new CMP0069 has much better LTO support for compiler other than intel. This was introduced in CMake 3.9.

**Comment by mhinz** [comment]
What would be the exact advantage for Neovim? Is there a benchmark that shows an improvement in any way?

Does anyone know how to check if Mach-O executable was compiled with LTO? (clang or gcc)

**Comment by jamessan** [comment]
> What would be the exact advantage for Neovim?

The same sorts of advantages that come from compiling with `-O2` rather than `-O0`, except that with LTO the compiler is able to make optimizations across translation units.

> Does anyone know how to check if Mach-O executable was compiled with LTO?

There's no marker for the compiler using LTO.  It's sort of doing another compilation pass but with all of the intermediate files available, which is why the linker needs the same optimization flags as the original compile.

**Comment by KillTheMule** [comment]
As an unrelated note, in my little rust project adding lto has had a pretty serious performance impact. Benchmarks would be best, of course.

**Comment by mhinz** [comment]
@jamessan 

Ah, okay. I read that gcc added a `.gnu.lto` section header to an ELF and hoped it (or clang) would do something similar for Mach-O. 

Best answer I found: https://stackoverflow.com/questions/51048414/clang-how-to-check-if-lto-was-performed

Well, LGTM.

**Comment by justinmk** [comment]
MINGW_64 build failed. Restarted.

**Comment by zhou13** [comment]
MINGW_64 build seems to be OK. Failed on test.

**Comment by justinmk** [comment]
Restarted AppVeyor again. It's very unusual for the AppVeyor build to fail, so I'm hesitant to merge this until it succeeds.

**Comment by ZviRackover** [comment]
Seems that Mingw32 passed while Mingw64 failed with a bunch of "plugin needed to handle lto object" errors. 
FWIW, this is a similar situation to the StackOverflow question https://stackoverflow.com/questions/32221221/mingw-x64-windows-plugin-needed-to-handle-lto-object which has a solution to the issue.

**Comment by zhou13** [comment]
@ZviRackover Thanks for finding the clues.

@justinmk Should I just disable LTO for mingw, or add put the LTO plugin into the bfd-plugins directory?  I don't think I can change the appveyor's CI.

**Comment by justinmk** [comment]
> Should I just disable LTO for mingw, 

Let's do that for now.

**Comment by zhou13** [comment]
Updated.  Waiting for the CI.

**Pull request merged**